kostal/__init__.py

```diff
@@ -128,72 +132,67 @@ def parse_item(self, item):
+            self.logger.error(
+                'could not retrieve data from {0}: {1}'.format(self.ip, e))
+            return
+
     def _refresh(self):
         start = time.time()
-        if self.datastucture == "html":
-            #HTML-OLD-Coding
-            try:
-                data = self._sh.tools.fetch_url(
-                    'http://' + self.ip + '/', self.user, self.passwd, timeout=2).decode()
-                # remove all attributes for easy findall()
-                data = re.sub(r'<([a-zA-Z0-9]+)(\s+[^>]*)>', r'<\1>', data)
-                # search all TD elements
-                table = re.findall(r'<td>([^<>]*)</td>', data, re.M | re.I | re.S)
-                for kostal_key in self._keytable:
-                    value = table[self._keytable[kostal_key]].strip()
-                    if 'x x x' not in value:
-                        self.logger.debug('set {0} = {1}'.format(kostal_key, value))
-                        if kostal_key in self._items:
-                            self._items[kostal_key](value)
-            except Exception as e:
-                self.logger.error(
-                    'could not retrieve data from {0}: {1}'.format(self.ip, e))
-                return
-        else:
-            #NEW-JSON-Coding
-            try:
-                kostalurl = 'http://' + self.ip + '/api/dxs.json?sessionid=SmartHomeNG'
-                for kostal_key in self._keytable:
-                    value = self._keytable[kostal_key]
-                    kostalurl +='&dxsEntries=' + str(value)
-                with urllib.request.urlopen(kostalurl) as url:
-                    data = json.loads(url.read().decode())
-                    for values in data['dxsEntries']:
-                        kostal_key = str(list(self._keytable.keys())[list(self._keytable.values()).index(values['dxsId'])])
-                        value=values['value']
-                        if kostal_key == "operation_status":
-                            self.logger.debug("operation_status" + str(value))
-                            if str(value) == "0":
-                                value = "off"
-                            elif str(value) == "2":
-                                value = "startup"
-                            elif str(value) == "3":
-                                value = "feed in (mpp)"
-                            elif str(value) == "6":
-                                value = "dc voltage low"
-                            else:
-                                value = "unknown"
-                        if kostal_key in self._items:
-                            self._items[kostal_key](value)
-                            self.logger.debug("items[" + str(kostal_key) +"] = " +str(value))
-            except Exception as e:
-                self.logger.error(
-                    'could not retrieve data from {0}: {1}'.format(self.ip, e))
-                return
+        # run the working methods
+        self.datastructure()
```

Comment:
General exception handling could be removed from the different datastructure metehods and moved to this place to have it only one time. But it's only sugar - so not really required.